torch/_functorch/autograd_function.py

```diff
@@ -274,6 +442,50 @@ def __setattr__(self, name, value):
             return
         return setattr(self._pt_inner_ctx, name, value)
 
+ctx_custom_save_properties = ('_pt_inner_ctx', '_pt_saved_tensors_bdims', '_pt_current_level')
+
+class CtxCustomSave:
+    def __init__(self, ctx, current_level):
+        if isinstance(ctx, torch.autograd.function.FunctionCtx):
+            for name in ctx_custom_save_properties:
+                if not hasattr(ctx, name):
+                    continue
+                raise RuntimeError(
+                    f'PyTorch reserves the {ctx_custom_save_properties} field on ctx. '
+                    'Please name your fields on ctx something else to avoid name '
+                    'collision.')
+        self._pt_inner_ctx = ctx
+        self._pt_saved_tensors_bdims = ()
+        self._pt_current_level = current_level
+
+    def save_for_backward(self, *tensors):
+        unwrapped_tensors, bdims = unwrap_batched(tensors, self._pt_current_level)
+        self._pt_inner_ctx.save_for_backward(*unwrapped_tensors)
+        self._pt_saved_tensors_bdims = bdims
+
+    def __getattr__(self, attr):
+        return getattr(self._pt_inner_ctx, attr)
+
+    def __setattr__(self, attr, value):
+        if attr in ctx_custom_save_properties:
+            self.__dict__[attr] = value
+            return
+        return setattr(self._pt_inner_ctx, attr, value)
```

Comment:
These ctx wrappers do different things so I would prefer to keep them separate. This one has custom save_for_backward behavior, the other one is used for overriding ctx.saved_tensors.

The similarity is in how they're both context wrappers that forward attr accesses to the underlying ctx object. I can make a common base class that both of these inherit from to avoid repetition, if that sounds reasonable

Reply:
Sounds good!

Reply:
Updated